File: kaji/sdk/kaji/runtime/integrations/functional.py
"""Function-level ``@tool`` for one-off tools without an ``Integration`` subclass.

The class-based ``Integration`` path remains the right shape for namespaced,
multi-tool bundles. This module adds a lightweight alternative for the common
case of "I have one async function; expose it as a tool."

    @tool
    async def get_weather(city: str) -> dict:
        return {"city": city, "tempF": 68}

    runtime = AgentBuilder().provider(p).tool(get_weather).build()

The handler's type hints are introspected to build a Pydantic model, which is
then converted to JSON Schema the same way ``@tool(parameters=Model)`` does.
"""

from __future__ import annotations

import inspect
import logging
from dataclasses import replace
from typing import Any, Callable, Dict, Optional, Type, Union, cast, get_type_hints

from pydantic import BaseModel, Field, create_model

from kaji.runtime.tools.registry import (
    ToolHandler,
    ToolRegistry,
    ToolSpec,
    provider_safe_tool_name,
    tool_spec_from_model,
)

logger = logging.getLogger(__name__)


def _warn_on_sanitize(original: str, sanitized: str) -> None:
    logger.warning(
        "tool name %r sanitized to %r for provider compatibility",
        original,
        sanitized,
    )


class BoundTool:
    """A single tool packaged with its spec and handler, registrable like an
    ``Integration``. Built by the ``@tool`` decorator; consumed by
    ``AgentBuilder.tool(...)``.
    """

    __slots__ = ("spec", "handler", "namespace")

    def __init__(
        self, spec: ToolSpec, handler: ToolHandler, namespace: str = "fn"
    ) -> None:
        self.spec = spec
        self.handler = handler
        self.namespace = namespace

    def register(self, registry: ToolRegistry) -> None:
        catalog_name = f"{self.namespace}.{self.spec.name}"
        prefixed = replace(
            self.spec,
            name=provider_safe_tool_name(catalog_name, on_mutate=_warn_on_sanitize),
            catalog_name=catalog_name,
        )
        registry.register(prefixed)(self.handler)


def _model_from_signature(fn: Callable[..., Any], model_name: str) -> Type[BaseModel]:
    """Build a Pydantic model from the handler's annotated parameters.

    Skips ``self`` (for methods) and ``ctx`` (the legacy ToolContext parameter).
    Raises ``TypeError`` if a parameter has no annotation; we fail loudly rather
    than guess.
    """
    sig = inspect.signature(fn)
    hints = get_type_hints(fn)
    fields: Dict[str, Any] = {}
    for name, param in sig.parameters.items():
        if name in ("self", "ctx"):
            continue
        if name not in hints:
            raise TypeError(
                f"@tool function {fn.__name__!r} parameter {name!r} has no type "
                "annotation; annotate it or pass parameters= explicitly."
            )
        default = param.default if param.default is not inspect.Parameter.empty else ...
        fields[name] = (hints[name], Field(default))
    return create_model(model_name, **fields)


def _wrap_handler(fn: Callable[..., Any]) -> ToolHandler:
    """Adapt ``fn(arg1, arg2, ...)`` into the ``(ctx, args_dict)`` shape the
    registry expects.

    The wrapper unpacks ``args_dict`` as keyword arguments and normalises the
    return value: non-dict results (primitives, lists, models) are wrapped as
    ``{"result": ...}`` so the tool log carries a JSON-serialisable object.
    """
    sig = inspect.signature(fn)
    param_names = [n for n in sig.parameters if n not in ("self", "ctx")]

    async def adapter(ctx: Any, args: Dict[str, Any]) -> Dict[str, Any]:
        kwargs = {name: args[name] for name in param_names if name in args}
        result = await fn(**kwargs)
        return result if isinstance(result, dict) else {"result": result}

    return adapter  # type: ignore[return-value]


def function_tool(
    _fn: Optional[Callable[..., Any]] = None,
    *,
    description: Optional[str] = None,
    parameters: Optional[Union[Dict[str, Any], Type[BaseModel]]] = None,
    risk: Optional[str] = None,
    tags: tuple[str, ...] = (),
    enabled: bool = True,
    namespace: str = "fn",
) -> Any:
    """Wrap a bare async function as a registrable tool.

    Usage::

        @function_tool
        async def get_weather(city: str) -> dict:
            return {"city": city, "tempF": 68}

        @function_tool(description="Look up weather.", risk="read")
        async def get_weather(city: str) -> dict: ...

    When ``parameters`` is omitted, the schema is derived from the function's
    type hints. Pass ``parameters=Model`` or a JSON Schema dict to override.
    """

    def make(fn: Callable[..., Any]) -> BoundTool:
        params_schema: Dict[str, Any]
        if parameters is None:
            model = _model_from_signature(fn, f"{fn.__name__}__Args")
            params_schema = tool_spec_from_model("_", "_", model).parameters
        elif isinstance(parameters, type) and issubclass(parameters, BaseModel):
            params_schema = tool_spec_from_model("_", "_", parameters).parameters
        else:
            params_schema = cast(Dict[str, Any], parameters)

        spec = ToolSpec(
            name=fn.__name__,
            description=description or fn.__doc__ or fn.__name__,
            parameters=params_schema,
            risk=risk,
            tags=tags,
            enabled=enabled,
        )
        return BoundTool(spec=spec, handler=_wrap_handler(fn), namespace=namespace)

    # Support both bare ``@tool`` and parameterised ``@tool(...)`` forms.
    if _fn is not None:
        return make(_fn)
    return make

File: kaji/packages/ts/src/integrations/functional.ts
/**
 * Function-level `functionTool` for one-off tools without an `Integration` subclass.
 * Mirrors `kaji.runtime.integrations.functional`.
 *
 * The class-based `Integration` path remains the right shape for namespaced,
 * multi-tool bundles. This module adds a lightweight alternative for the common
 * case of "I have one async function; expose it as a tool."
 *
 *   const getWeather = functionTool(
 *     { description: "Return weather.", parameters: z.object({ city: z.string() }) },
 *     async ({ city }) => ({ city, tempF: 68 }),
 *   );
 *
 *   const runtime = new AgentBuilder().provider(p).tool(getWeather).build();
 */
import * as z from "zod";

import {
  TOOL_ARGUMENT_VALIDATOR,
  providerSafeToolName,
  setToolArgumentValidator,
  toolArgumentValidator,
  type ToolHandler,
  type ToolMeta,
  type ToolRegistry,
  type ToolSpec,
  toolParametersToJSONSchema,
} from "@/tools/registry";
import type { ToolExecutionContext } from "@/runtime/context";
import {
  cloneToolExecutionArguments,
  consumeValidationReceipt,
  validateIsolatedToolArguments,
} from "@/tools/validation";

type ArgsOf<P> = P extends z.ZodType ? z.input<P> : Record<string, unknown>;

/** Signature accepted by `functionTool`. Zod validates the provider arguments,
 * but its defaults, coercions, and transformations are deliberately discarded.
 * The second context parameter may be ignored and matches registry handlers. */
export type FunctionToolHandler<P> = (
  args: ArgsOf<P>,
  context: ToolExecutionContext,
) => Promise<unknown>;

/** Tool packaged with its spec + adapter handler, registrable like an Integration. */
export class BoundTool {
  constructor(
    readonly spec: ToolSpec,
    readonly handler: ToolHandler,
    readonly namespace: string = "fn",
  ) {}

  register(registry: ToolRegistry): void {
    const catalogName = `${this.namespace}.${this.spec.name}`;
    registry.register(
      setToolArgumentValidator(
        {
          ...this.spec,
          name: providerSafeToolName(catalogName, { onMutate: warnOnSanitize }),
          catalogName,
        },
        this.spec[TOOL_ARGUMENT_VALIDATOR],
      ),
      this.handler,
    );
  }
}

function warnOnSanitize(original: string, sanitized: string): void {
  console.warn(
    `[kaji] tool name ${JSON.stringify(original)} sanitized to ${JSON.stringify(sanitized)} for provider compatibility`,
  );
}

export interface FunctionToolMeta<P> extends Omit<ToolMeta, "parameters"> {
  /** Tool name. Defaults to the handler's `.name` if available. */
  name?: string;
  /** Zod schema or JSON Schema describing the arguments. */
  parameters: P;
  /** Namespace prefix (defaults to "fn"). */
  namespace?: string;
}

/**
 * Build a single tool from a meta + handler pair. The handler receives the
 * a byte-equivalent isolated clone after validation succeeds.
 */
export function functionTool<P extends z.ZodType | Record<string, unknown>>(
  meta: FunctionToolMeta<P>,
  handler: FunctionToolHandler<P>,
): BoundTool {
  const name = meta.name ?? handler.name ?? "tool";
  const argumentValidator = toolArgumentValidator(meta.parameters as never);
  const spec = setToolArgumentValidator<ToolSpec>(
    {
      name,
      description: meta.description,
      parameters: toolParametersToJSONSchema(meta.parameters as never),
      risk: meta.risk,
      ...(meta.tags !== undefined ? { tags: meta.tags } : {}),
      ...(meta.enabled !== undefined ? { enabled: meta.enabled } : {}),
      ...(meta.parallel_safe !== undefined ? { parallel_safe: meta.parallel_safe } : {}),
      ...(meta.timeout_ms !== undefined ? { timeout_ms: meta.timeout_ms } : {}),
    },
    argumentValidator,
  );

  const adapter: ToolHandler = async (args, context) => {
    let executionArgs = args;
    if (
      argumentValidator !== undefined &&
      !consumeValidationReceipt(context, args, argumentValidator)
    ) {
      executionArgs = cloneToolExecutionArguments(name, args);
      await validateIsolatedToolArguments(name, executionArgs, argumentValidator);
    }
    const result = await handler(executionArgs as ArgsOf<P>, context);
    // ToolHandler must return an object; wrap primitives/arrays.
    if (result !== null && typeof result === "object" && !Array.isArray(result)) {
      return result as Record<string, unknown>;
    }
    return { result };
  };
  return new BoundTool(spec, adapter, meta.namespace ?? "fn");
}
